presto-main/src/main/java/com/facebook/presto/tdigest/TDigest.java

```diff
@@ -0,0 +1,664 @@
+            sliceInput.close();
+            return r;
+        }
+        catch (IndexOutOfBoundsException e) {
+            throw new IllegalArgumentException("Incorrect slice serialization format");
+        }
+    }
+
+    public void add(double x)
+    {
+        add(x, 1);
+    }
+
+    public void add(double x, long w)
+    {
+        checkArgument(!Double.isNaN(x), "Cannot add NaN to t-digest");
+        checkArgument(w > 0L, "weight must be > 0");
+
+        if (tempUsed >= tempWeight.length - activeCentroids - 1) {
+            mergeNewValues();
+        }
+        int where = tempUsed++;
+        tempWeight[where] = w;
+        tempMean[where] = x;
+        unmergedWeight += w;
+        if (x < min) {
+            min = x;
+        }
+        if (x > max) {
+            max = x;
+        }
+    }
+
+    public void merge(TDigest other)
+    {
+        checkArgument(other != null, "Cannot merge with a null t-digest");
+        checkArgument(this.publicCompression == other.getCompressionFactor(), format("TDigests must have the same compression, found (%s, %s)", this.publicCompression,
+                other.getCompressionFactor()));
+        List<Centroid> tmp = new ArrayList<>();
+        for (Centroid centroid : other.centroids()) {
+            tmp.add(centroid);
+        }
+
+        shuffle(tmp, gen);
+        for (Centroid centroid : tmp) {
+            add(centroid.getMean(), centroid.getWeight());
+        }
+    }
+
+    private void mergeNewValues()
+    {
+        mergeNewValues(false, compression);
+    }
+
+    private void mergeNewValues(boolean force, double compression)
+    {
+        if (unmergedWeight == 0) {
+            return;
+        }
+        if (force || unmergedWeight > 0) {
+            // note that we run the merge in reverse every other merge to avoid left-to-right bias in merging
+            merge(tempMean, tempWeight, tempUsed, order, unmergedWeight,
+                    mergeCount % 2 == 1, compression);
+            mergeCount++;
+            tempUsed = 0;
+            unmergedWeight = 0;
+        }
+    }
+
+    private void merge(double[] incomingMean, double[] incomingWeight, int incomingCount, int[] incomingOrder,
+            double unmergedWeight, boolean runBackwards, double compression)
+    {
+        System.arraycopy(mean, 0, incomingMean, incomingCount, activeCentroids);
+        System.arraycopy(weight, 0, incomingWeight, incomingCount, activeCentroids);
+        incomingCount += activeCentroids;
+
+        if (incomingOrder == null) {
+            incomingOrder = new int[incomingCount];
+        }
+        sort(incomingOrder, incomingMean, incomingCount);
+
+        if (runBackwards) {
+            reverse(incomingOrder, 0, incomingCount);
+        }
+
+        totalWeight += unmergedWeight;
+
+        checkArgument((activeCentroids + incomingCount) > 0, format("Active centroids plus incoming count must be > 0, was %s", activeCentroids + incomingCount));
```

Comment:
nit:
- `(activeCentroids + incomingCount) > 0` -> `activeCentroids + incomingCount > 0`
- remove `format`; it's redundant